### ISSUE TITLE
onInitialized gets called first instead of creationFunction if using injectFuture/injectStream

### ISSUE THREAD

**Issue by isanimu** [GIfatahTH/states_rebuilder#155]
**states_rebuilder v3.2.0**

```dart
enum AuthStatus { unknown, unauthenticated, guest, authenticated }
```

Expected Behavior:
```dart
final authStatus = RM.inject<AuthStatus>(
  () {
    return AuthStatus.unknown; // called first
  },
  onInitialized: (state) {
    print(state); // called after creation
  },
);

// inside build
authStatus.rebuilder() // gets called with state == AuthStatus.unknown
```
Unexpected Behavior:
```dart
final authStatus = RM.injectStream<AuthStatus>(
  () async* {
    yield AuthStatus.unknown; // called after onInitialized
  },
  onInitialized: (state) {
    print(state); // called first before creation, state == null
  },
);

// inside build
authStatus.rebuilder() // called with state == null
```

The only way to have state != null is by setting the **intialValue** property. I'm aware that what I was called the 'unexpected behavior' might be intended by this library.

But I wish there was something like onCreated which instead of giving state as the argument, it's giving the injected instance so then I could call setState to trigger onData/rebuild.

Let say I have a presenter layer as the state, Dart doesn't support async constructor and cascade await, but I could call something like scheduleMicrotask and have async callback on there, then onCreated I could call the setState to trigger rebuild.

I want to avoid something like,
```dart
class AuthPresenter {
  AuthStatus status;

  void initialize() async {
    status = await repository.getStatus();
  }
}

final authPresenter = RM.injectFuture<AuthPresenter>(
  () async {
    // unnecessary boilerplate
    final presenter = AuthPresenter();
    await presenter.initialize();
    return presenter;
  },
  onData: (presenter) {
    print(presenter.status);
  },
);
```

Instead I would love to have something like,
```dart
class AuthPresenter {
  AuthStatus status;

  AuthPresenter() {
    scheduleMicrotask(initialize);
  }

  void initialize() async {
    status = await locator<UserRepository>.getStatus();
  }
}

final authPresenter = RM.inject<AuthPresenter>(
  () => AuthPresenter(),
  onCreated: (injected) {
    injected.rebuild(); // instead of injected.setState((){});
  },
  onData: (presenter) {
    print(presenter.status);
  },
);
```
This shouldn't break the library.

1. OnCreated will always be called after creation & onInitialized
2. The argument should be the injected instance instead of the state.
3. Injected should have something like 'rebuild' function for triggering rebuild without changing the state, I mean something like injected.setState((){}) is the same as triggering rebuild without doing any state changes.

This sounds like an anti-pattern, and rebuild on onCreated already looked like a boilerplate itself that devs must have deal with async initialization. But I'm ok with that.

Basically by having the same behavior on onCreated, and having injected as the argument instead of the state, devs can do any kind of initialization whether it's a sync or async.

**Comment by isanimu** [comment]
Also one could create something like,

```dart
final Injected<AuthPresenter> authPresenter = RM.inject(
  () => AuthPresenter(),
  onInitialized: (_) {
    authPresenter.setState((){}); // works, but bad 
  },
  onData: (presenter) {
    authPresenter.setState((){}); // stack overflow
    print(presenter.status);
  },
);
```
This is way I'm proposing the above solution.

**Comment by isanimu** [comment]
Currently this is what I did to reduce the boilerplate

```dart
abstract class Initializable {
  void initialize();
}

Future<T> initializer<T extends Initializable>(T target) async {
  await target.initialize();
  return target;
}

class AuthPresenter implements Initializable {
  AuthStatus status;

  void initialize() async {
    status = await locator.get<UserRepository>().getStatus();
  }
}

final authPresenter = RM.injectFuture<AuthPresenter>(
  () async => await initializer(AuthPresenter()),
  onData: (presenter) {
    print(presenter.status);
  },
);
```

But this still doesn't solve the problem with the state is being null without having to set the initialValue property first.

**Comment by isanimu** [comment]
Maybe also adding a new bool property like buildAfterCreated, so the rebuilder will only gets called after onCreated. This way initialValue could still be null and rebuilder doesn't have to deal with null state.

**Comment by GIfatahTH** [comment]
@isanimu 
>The only way to have state != null is by setting the intialValue property. I'm aware that what I was called the 'unexpected behavior' might be intended by this library.

Yes it is the expected behavior. Streams when first fired and if they have no initial value, they are in the waiting state with null value.


>But I wish there was something like onCreated which instead of giving state as the argument, it's giving the injected instance so then I could call setState to trigger onData/rebuild

`onInitialized `does exactly what `onCreated `does as you explain it. For the argument, you can use the injected instance inside the body of the `onInitialized` :
```dart
  onInitialized: (_) {
    authPresenter.setState((){}); // works, but bad 
  },
```

>  () async {
    // unnecessary boilerplate
    final presenter = AuthPresenter();
    await presenter.initialize();
    return presenter;
  },

Why don't you sue the spread notation? (double dots)
```
final authPresenter = RM.injectFuture<AuthPresenter>(
  () => AuthPresenter()..initialize(),
  onData: (presenter) {
    print(presenter.status);
  },
);
```


>This way initialValue could still be null and rebuilder doesn't have to deal with null state.

In the next update of the library, and with null safety, state can not be null, and the initial value will be inferred by the library for primitives, otherwise, it must be defined explicitaly.

**Comment by isanimu** [comment]
@GIfatahTH Thanks for the answer.

I'll close the issue.

BTW, I was just wondering why don't make the Injected class non abstract?

Instead of making extension on the state class directly,

```dart
class MyModel {
  int counter;
}

class MyState {
   final model;
  void increase() {
    model.counter += 1;
  }
}
```
which force to access it through the state object,
```dart
injected.state.increase();
```

If only the Injected class is non abstract, one can make something like:
```dart
extension on Injected<MyPureState> {
  void increase() {
    state += 1;
  }
}
```
Thus will make this possible, and there is no need for multiple layer. The state can be a pure state, the injected can be a direct presenter/interactor.
```dart
injected.increase();
```

**Comment by amoslai5128** [comment]
> @GIfatahTH Thanks for the answer.
> 
> I'll close the issue.
> 
> BTW, I was just wondering why don't make the Injected class non abstract?
> 
> Instead of making extension on the state class directly,
> 
> ```dart
> class MyModel {
>   int counter;
> }
> 
> class MyState {
>    final model;
>   void increase() {
>     model.counter += 1;
>   }
> }
> ```
> 
> which force to access it through the state property,
> 
> ```dart
> injected.state.increase();
> ```
> 
> If only the Injected class is non abstract, one can make something like:
> 
> ```dart
> extension on Injected<MyPureState> {
>   void increase() {
>     state += 1;
>   }
> }
> ```
> 
> Thus will make this possible, and there is no need for multiple layer. The state can be a pure state, the injected can be a direct presenter/interactor.
> 
> ```dart
> injected.increase();
> ```

It did use the extension once at the [example 009](https://github.com/GIfatahTH/states_rebuilder/tree/master/examples/ex_009_1_4_ca_todo_mvc_with_state_persistence_and_user_auth)

I agree this would make the code to be more developer friendly.

**Comment by isanimu** [comment]
It turns out that Dart can create extension on abstract class. OK, never mind.

Thanks.
